[libc++] Move some macOS CI jobs to Github actions

ldionne opened the pull request:
This patch decouples macOS CI testing from BuildKite, which makes the
maintenance of macOS CI easier and more accessible to all contributors.
Right now, the macOS CI is running entirely on machines owned by the
LLVM Foundation with only a small set of contributors having direct
access to them. In particular, updating these machines is currently
a very time-consuming manual process that requires taking the machines
offline, and using Github-provided instances makes that an order of
magnitude easier.

The story for performing back-deployment testing still needs to be
figured out, so for now we are retaining some jobs under BuildKite.

EricWF commented:
Let's wait for the results to come back, but this LGTM.
Are we using the Github hosted machines for these builds now? If so, we probably need permission from the LLVM org before we start eating minutes.

tstellar commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -199,6 +201,38 @@ jobs:
             **/crash_diagnostics/*
+
+  macos:
+    runs-on: macos-latest
```
I recommend using macos-14 here, so we can control when we update to the next macos version.

tstellar reviewed (commented):
How often will this job be running?  We are limited to only 5 concurrent macOS jobs at a time, so if this is going to be running for each PR, we may quickly run out of capacity.

boomanaiden154 commented:
If everything is already running on machines owned by the foundation, it should just be a matter of setting up the Github actions agent on them to get more runner availability? Like @tstellar mentioned, that's not what this PR does currently though.

ldionne commented:
> How often will this job be running? We are limited to only 5 concurrent macOS jobs at a time, so if this is going to be running for each PR, we may quickly run out of capacity.

This would run after stage1 has completed. Only a subset of all PRs make it there, as most will fail in stage1.

> If everything is already running on machines owned by the foundation, it should just be a matter of setting up the Github actions agent on them to get more runner availability? Like @tstellar mentioned, that's not what this PR does currently though.

Yes, we could do that or we could pay for additional capacity on Github directly. I have been maintaining the Foundation macOS nodes and not having to do that anymore is the main motivation for this change.

tstellar commented:
> Yes, we could do that or we could pay for additional capacity on Github directly. I have been maintaining the Foundation macOS nodes and not having to do that anymore is the main motivation for this change.

Would we be able to repurpose those machines into self-hosted GitHub runners?  If so, I can help with the maintenance of those machines.  That would help ensure we have the capacity we need.

ldionne commented:
> > Yes, we could do that or we could pay for additional capacity on Github directly. I have been maintaining the Foundation macOS nodes and not having to do that anymore is the main motivation for this change.
> 
> Would we be able to repurpose those machines into self-hosted GitHub runners? If so, I can help with the maintenance of those machines. That would help ensure we have the capacity we need.

Yes, that should be possible. I will start by making it work for the Github-provided builders and then we can tackle the capacity increase as a separate task.

boomanaiden154 commented:
@ldionne This PR seems to have a lot of unrelated libc++ changes that somehow got pulled in?

Also, we'll have to think about the capacity for this. I believe we can only have five concurrent macOS jobs running at the same time. I'm not sure how many concurrent jobs the libc++ CI normally runs or how long the macOS jobs run for, but this could quickly end up becoming an issue. I don't think it would block too much else in the project though as everything (that I can think of) runs on Linux runners, other than the release jobs.

ldionne commented:
> @ldionne This PR seems to have a lot of unrelated libc++ changes that somehow got pulled in?

I am just using this to iterate on fixing flaky tests that only show up on the Github-provided macOS builders. These changes won't be part of the final patch.

> 
> Also, we'll have to think about the capacity for this. I believe we can only have five concurrent macOS jobs running at the same time. I'm not sure how many concurrent jobs the libc++ CI normally runs or how long the macOS jobs run for, but this could quickly end up becoming an issue. I don't think it would block too much else in the project though as everything (that I can think of) runs on Linux runners, other than the release jobs.

I agree. We'd probably want to set up the existing Foundation-provided macOS machines as runners available via Github actions, but I wanted to look into that separately.

ldionne commented:
@philnik777 Across all jobs, it looks like the following tests are failing:

```
std/thread/thread.mutex/thread.lock/thread.lock.unique/thread.lock.unique.cons/mutex_duration.pass.cpp
std/thread/thread.mutex/thread.lock/thread.lock.unique/thread.lock.unique.cons/mutex_time_point.pass.cpp
std/thread/thread.mutex/thread.lock/thread.lock.unique/thread.lock.unique.cons/mutex.pass.cpp
std/thread/thread.mutex/thread.lock/thread.lock.unique/thread.lock.unique.locking/lock.pass.cpp
std/thread/thread.mutex/thread.mutex.requirements/thread.mutex.requirements.mutex/thread.mutex.class/lock.pass.cpp
std/thread/thread.mutex/thread.mutex.requirements/thread.shared_mutex.requirements/thread.shared_mutex.class/lock_shared.pass.cpp
std/thread/thread.mutex/thread.mutex.requirements/thread.shared_mutex.requirements/thread.shared_mutex.class/lock.pass.cpp
std/thread/thread.mutex/thread.mutex.requirements/thread.sharedtimedmutex.requirements/thread.sharedtimedmutex.class/try_lock_for.pass.cpp
std/thread/thread.mutex/thread.mutex.requirements/thread.sharedtimedmutex.requirements/thread.sharedtimedmutex.class/try_lock_shared_for.pass.cpp
std/thread/thread.mutex/thread.mutex.requirements/thread.sharedtimedmutex.requirements/thread.sharedtimedmutex.class/try_lock_until.pass.cpp
std/thread/thread.mutex/thread.mutex.requirements/thread.timedmutex.requirements/thread.timedmutex.class/lock.pass.cpp
std/thread/thread.mutex/thread.mutex.requirements/thread.timedmutex.requirements/thread.timedmutex.class/try_lock_for.pass.cpp
std/thread/thread.mutex/thread.mutex.requirements/thread.timedmutex.requirements/thread.timedmutex.class/try_lock_until.pass.cpp
std/thread/thread.mutex/thread.mutex.requirements/thread.timedmutex.requirements/thread.timedmutex.recursive/lock.pass.cpp
std/thread/thread.mutex/thread.mutex.requirements/thread.timedmutex.requirements/thread.timedmutex.recursive/try_lock_for.pass.cpp
std/thread/thread.mutex/thread.mutex.requirements/thread.timedmutex.requirements/thread.timedmutex.recursive/try_lock_until.pass.cpp
```

ldionne committed:
[libc++] Move some macOS CI jobs to Github actions

This patch decouples macOS CI testing from BuildKite, which makes the
maintenance of macOS CI easier and more accessible to all contributors.
Right now, the macOS CI is running entirely on machines owned by the
LLVM Foundation with only a small set of contributors having direct
access to them. In particular, updating these machines is currently
a very time-consuming manual process that requires taking the machines
offline, and using Github-provided instances makes that an order of
magnitude easier.

The story for performing back-deployment testing still needs to be
figured out, so for now we are retaining some jobs under BuildKite.

ldionne commented:
I looked at the policy for Github Actions runner minutes here: https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions#about-billing-for-github-actions. For public repositories, there is no limit of minutes per month. There is a maximum of 5 macOS runners at a time which could be a problem for capacity if we want to start spinning a lot more macOS CI jobs. However, I would cross that bridge when we get to it.

In particular, the goal of this transition is to move away from the MacMiniVault macOS runners we're currently using for BuildKite because they are very tedious to keep up-to-date. I'd like to avoid setting up and maintaining Github-hosted runners on those machines since that would defeat the main benefit of moving to Github actions.

@tstellar Are you fine with me landing this patch and then figuring out how to best address capacity issues if we have them? In the short term, one benefit of landing this patch is that we could eventually let go some of the machines we have in MacMiniVault (I think we can let go the two arm64 machines).

tstellar commented:
Let's try it out and see how it goes.

firewave commented:
FYI the performance of `macos-*` runners can be quite erratic and they may be performing multitudes (up to 5x) worse on occasion - see https://github.com/danmar/cppcheck/pull/6704 for some context. I guess I also need to report this upstream.
